harvester_e2e_tests/integration/test_upgrade.py

```diff
@@ -1,121 +1,1217 @@
         """
         - perform upgrade
         - check all nodes upgraded
         """
+        version = request.config.getoption("--upgrade-target-version") or "master"
+        assert not _is_installed_version(api_client, version), (
+            f"The current version is already {version}")
+
+        # force create upgrade version
+        code, data = api_client.versions.get(version)
+        if code == 200:
+            code, data = api_client.versions.delete(version)
+            assert code == 204, (
+                f"Failed to delete version {version}: {data}")
+
+        code, data = _create_version(request, api_client, version)
+        assert code == 201, (
+            f"Failed to create version {version}: {data}")
+
+        code, data = api_client.upgrades.create(version)
+        assert code == 201, (
+            f"Failed to upgrade version {version}: {code}, {data}")
+
+        def _wait_for_upgrade():
+            try:
+                code, upgrade_data = api_client.upgrades.get(data["metadata"]["name"])
+                if code != 200:
+                    return False
+            except Exception:
+                return False
+
+            if upgrade_data['metadata'].get('labels', {}).get(UPGRADE_STATE_LABEL) == "Succeeded":
+                return True
+
+            conds = upgrade_data.get('status', {}).get('conditions', [])
+            if len(conds) > 0:
+                for cond in conds:
+                    if cond["status"] == "False":
+                        cond_type = cond["type"]
+                        raise AssertionError(f"Upgrade failed: {cond_type}: {cond}")
+
+                    if cond["type"] == "Completed" and cond["status"] == "True":
+                        return True
+
+            return False
 
-    @pytest.mark.dependency(depends=["three_nodes_upgrade"])
-    def test_verify_logging(self):
+        upgrade_timeout = request.config.getoption('--upgrade-wait-timeout') or 7200
+        endtime = datetime.now() + timedelta(seconds=upgrade_timeout)
+        while endtime > datetime.now():
+            if _wait_for_upgrade():
+                break
+            sleep(5)
+        else:
+            raise AssertionError("Upgrade timeout")
+
+        # start vms when are stopped
+        code, data = api_client.vms.get()
+        assert code == 200, (f"Failed to get vms: {data}")
+
+        for vm in data["data"]:
+            if "ready" not in vm["status"] or not vm["status"]["ready"]:
+                code, data = api_client.vms.start(vm["metadata"]["name"])
+                assert code == 204, (f"Failed to start vm {vm['metadata']['name']}: {data}")
+
+                # wait for vm to be assigned an IP
+                endtime = datetime.now() + timedelta(seconds=wait_timeout)
+                while endtime > datetime.now():
+                    if _wait_for_vm_ready(request, api_client, vm["metadata"]["name"],
```

Comment:
the helper function have only 3 arguments, It looks `request` should not be passed in here.